enhancements/20240820-mtu-and-network-information-propagation.md

```diff
@@ -0,0 +1,318 @@
+
+The related note needs to be added to Harvester document.
+
+### User Experience In Detail
+
+### API changes
+
+## Design
+
+### Implementation Overview
+
+#### Update NAD MTU on the fly
+
+No matter Harvester puts MTU on `Clusternetwork` or `Vlanconfig`, an enhancement is needed to:
+
+(1) Sync the information from Harvester to NAD
+
+(2) Allow NAD to have a different MTU than Harvester side.
+
+e.g. for test
+
+(3) Automatically replace/migrate the POD/VM to get the updated MTU, or warn the user that the change will only take effect after the POD/VM is replaced/rebooted.
+
+#### MTU belonging
+
+##### Solution 1: Keep on `Vlanconfig`
+
+Advantages:
+
+(i) MTU is a concept of physical nics, the `Vlanconfig` has `uplink` and the `MTU` is by nature adhered to it.
+
+(ii) More convenient for network adjustment.
+
+e.g. User plans to set a bigger MTU for the VM network, with the `Vlanconfig`, user can 
+
+```
+
+(a) Migrate/stop VM from the nodes belonging to Vlanconfig 1
+
+(b) Set a new MTU on those nodes
+
+(c) Create a few test VMs to test the new network
+
+(d.1) Migrate VM back
+
+(d.2) If not working as expected, revert the changes
+
+```
+Repeat (a)..(d) to finish the maintenance of a whole cluster.
+
+Also refer [Replace Ethernet NICs](https://docs.harvesterhci.io/v1.3/networking/best-practice#replace-ethernet-nics), the current `Vlanconfig` proves to be good for changing the network setting on the fly.
+
+(iii) `Vlanconfig` is the right place to cover/encapsulate `NAD` and keep evolving.
+
+Disadvantages:
+
+(i) A bit tedious to adjust the MTU on the fly.
+
+(ii) MTU inconsistency in the middle. It is also possible that the MTU was misconfigured to be not unique for the `Vlanconfig`s under the same `Clusternetwork`.
+
+When this feature [add Harvester alerts](https://github.com/harvester/harvester/issues/2850) is implemented, users will be notified from Harvester UI and the Harvester [Monitoring](https://docs.harvesterhci.io/v1.3/monitoring/harvester-monitoring#alertmanager).
+
+##### Solution 2: Move to `Clusternetwork`
+
+Advantages:
+
+(i) Ideally one click.
+
+Disadvantages:
+
+(i) `Clusternetwork` is like a virtual class, not fit for putting some fields here.
+
+(ii) When `MTU` is on `Clusternetwork`, it is still hard to drop `MTU` from `Vlanconfig`, refer above scenario of changing the MTU on a running cluster. If we have 3 places of MTU, the sync will be even more complex.
+
+(iii) When something is wrong, it will afffect all.
+
+#### The planned VxLan Feature
+
+`Clusternetwork` may still keep unchanged, and a new CRD like `Vxlanconfig` is added to act the similar role of `Vlanconfig`.
+
+From this point, `MTU` is not something belonging to the `Clusternetwork`.
+
+### Test plan
+
+Integration test plan.
+
+### Upgrade strategy
+
+Anything that requires if user want to upgrade to this enhancement
+
+## Note [optional]
+
+Additional notes.
```

Comment:
Generic Question:
we select only particular cluster network when creating vlan network.
In that case if a cluster network has more than one vlanconfig, which vlan config will be used for the vlan vm network?
My understanding is all vlanconfig/multiple physical NICs under a cluster network will connect to the same bridge and forward traffic in the fashion we select in bond options.
In that case is it really benefit table to have mtu field under vlanconfig ? Do we need to make sure it is same mtu configured for all vlan configs under the same cluster ? If that is the case, is it better to move the config under cluster network ?

Reply:
This is a good question.

From the design of current network implementation, a general assumption is that: 

**The physical nics on each node could be none-unified.** for none-management network, different nodes may join different `clusternetwork`. e.g., the later added feature `witness-node`, such node will not join any other network in general, it mainly gives a running location for `etcd`.

When creating a `Vlanconfig` (https://docs.harvesterhci.io/v1.3/networking/index#custom-cluster-network step 3), users needs to select which node(s) are the target. and the related uplink parameters.

![image](https://github.com/user-attachments/assets/5bd337c2-ca7a-4f04-b0d1-a97ae4e2bd1d)

Another example:

After a cluster runs some time, user adds some new servers with GPU and new nics, he can create a new cluster network and related vlanconfig for those server to have new services.

//
To add check and alarm of potential params differences like MTU is what we need to enhance.

Reply:
From recent practices, one `Vlanconfig` for one node / a group of node is also better for system maintenance.

Customer had operations like below on the fly.

https://docs.harvesterhci.io/v1.3/networking/best-practice#replace-ethernet-nics